File: control-contest-02-prepare/H/cpp/H.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <array>
#include <algorithm>

using namespace std;

using ll = long long;

const ll INF = 4e18;

struct Road {
    int to;
    int time;
};

struct State {
    ll time;
    int playground;
    int usedMismatch;

    bool operator>(const State& other) const {
        return time > other.time;
    }
};

vector<array<ll, 2>> dijkstra(
    int n,
    const vector<vector<Road>>& graph,
    int start,
    const vector<bool>& hasMismatch
) {
    vector<array<ll, 2>> dist(n + 1, {INF, INF});

    int startState = hasMismatch[start] ? 1 : 0;
    dist[start][startState] = 0;

    priority_queue<State, vector<State>, greater<State>> queue;
    queue.push({0, start, startState});

    while (!queue.empty()) {
        State current = queue.top();
        queue.pop();

        if (current.time != dist[current.playground][current.usedMismatch]) {
            continue;
        }

        for (const Road& road : graph[current.playground]) {
            int nextUsedMismatch = (current.usedMismatch || hasMismatch[road.to]) ? 1 : 0;
            ll newTime = current.time + (current.usedMismatch ? road.time / 2 : road.time);

            if (newTime < dist[road.to][nextUsedMismatch]) {
                dist[road.to][nextUsedMismatch] = newTime;
                queue.push({newTime, road.to, nextUsedMismatch});
            }
        }
    }

    return dist;
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int tests;
    cin >> tests;

    while (tests--) {
        int n, m, k;
        cin >> n >> m >> k;

        vector<bool> hasMismatch(n + 1, false);

        for (int i = 0; i < k; i++) {
            int playground;
            cin >> playground;
            hasMismatch[playground] = true;
        }

        vector<vector<Road>> graph(n + 1);

        for (int i = 0; i < m; i++) {
            int firstPlayground, secondPlayground, time;
            cin >> firstPlayground >> secondPlayground >> time;

            graph[firstPlayground].push_back({secondPlayground, time});
            graph[secondPlayground].push_back({firstPlayground, time});
        }

        vector<array<ll, 2>> madinDist = dijkstra(n, graph, 1, hasMismatch);
        vector<array<ll, 2>> goshaDist = dijkstra(n, graph, n, hasMismatch);

        ll answer = INF;

        for (int playground = 1; playground <= n; playground++) {
            ll madinTime = min(madinDist[playground][0], madinDist[playground][1]);
            ll goshaTime = min(goshaDist[playground][0], goshaDist[playground][1]);

            answer = min(answer, max(madinTime, goshaTime));
        }

        cout << (answer == INF ? -1 : answer) << '\n';
    }

    return 0;
}

File: control-contest-02-prepare/H/java/H.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.PriorityQueue;

public class H {
    static final long INF = Long.MAX_VALUE / 4;

    static class Road {
        int to;
        int time;

        Road(int to, int time) {
            this.to = to;
            this.time = time;
        }
    }

    static class State implements Comparable<State> {
        long time;
        int playground;
        int usedMismatch;

        State(long time, int playground, int usedMismatch) {
            this.time = time;
            this.playground = playground;
            this.usedMismatch = usedMismatch;
        }

        @Override
        public int compareTo(State other) {
            return Long.compare(this.time, other.time);
        }
    }

    static long[][] dijkstra(int n, ArrayList<Road>[] graph, int start, boolean[] hasMismatch) {
        long[][] dist = new long[n + 1][2];

        for (int i = 1; i <= n; i++) {
            Arrays.fill(dist[i], INF);
        }

        int startState = hasMismatch[start] ? 1 : 0;
        dist[start][startState] = 0;

        PriorityQueue<State> queue = new PriorityQueue<>();
        queue.add(new State(0, start, startState));

        while (!queue.isEmpty()) {
            State current = queue.poll();

            if (current.time != dist[current.playground][current.usedMismatch]) {
                continue;
            }

            for (Road road : graph[current.playground]) {
                int nextUsedMismatch = (current.usedMismatch == 1 || hasMismatch[road.to]) ? 1 : 0;
                long newTime = current.time + (current.usedMismatch == 1 ? road.time / 2 : road.time);

                if (newTime < dist[road.to][nextUsedMismatch]) {
                    dist[road.to][nextUsedMismatch] = newTime;
                    queue.add(new State(newTime, road.to, nextUsedMismatch));
                }
            }
        }

        return dist;
    }

    public static void main(String[] args) throws Exception {
        BufferedReader in = new BufferedReader(new InputStreamReader(System.in));

        int tests = Integer.parseInt(in.readLine());
        StringBuilder output = new StringBuilder();

        for (int test = 0; test < tests; test++) {
            String[] first = in.readLine().split(" ");

            int n = Integer.parseInt(first[0]);
            int m = Integer.parseInt(first[1]);
            int k = Integer.parseInt(first[2]);

            boolean[] hasMismatch = new boolean[n + 1];

            String[] mismatchPlaygrounds = in.readLine().split(" ");
            for (int i = 0; i < k; i++) {
                int playground = Integer.parseInt(mismatchPlaygrounds[i]);
                hasMismatch[playground] = true;
            }

            ArrayList<Road>[] graph = new ArrayList[n + 1];

            for (int i = 1; i <= n; i++) {
                graph[i] = new ArrayList<>();
            }

            for (int i = 0; i < m; i++) {
                String[] parts = in.readLine().split(" ");

                int firstPlayground = Integer.parseInt(parts[0]);
                int secondPlayground = Integer.parseInt(parts[1]);
                int time = Integer.parseInt(parts[2]);

                graph[firstPlayground].add(new Road(secondPlayground, time));
                graph[secondPlayground].add(new Road(firstPlayground, time));
            }

            long[][] madinDist = dijkstra(n, graph, 1, hasMismatch);
            long[][] goshaDist = dijkstra(n, graph, n, hasMismatch);

            long answer = INF;

            for (int playground = 1; playground <= n; playground++) {
                long madinTime = Math.min(madinDist[playground][0], madinDist[playground][1]);
                long goshaTime = Math.min(goshaDist[playground][0], goshaDist[playground][1]);

                answer = Math.min(answer, Math.max(madinTime, goshaTime));
            }

            output.append(answer == INF ? -1 : answer).append('\n');
        }

        System.out.print(output);
    }
}